chrome/browser/net/chrome_network_delegate.cc

```diff
@@ -327,6 +330,98 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
   return rv;
 }
 
+std::string generateBlockedElemhideJavaScript(std::string url, std::string relative_url) {
+  std::stringstream ss;
+
+  // basically copy-pasting JS from
+  // https://github.com/adblockplus/adblockpluschrome/blob/master/include.preload.js#L546
+  ss <<
+"function hideElement(element)\n\
+{\n\
+  function doHide()\n\
+  {\n\
+    let propertyName = \"display\";\n\
+    let propertyValue = \"none\";\n\
+    if (element.localName == \"frame\")\n\
+    {\n\
+      propertyName = \"visibility\";\n\
+      propertyValue = \"hidden\";\n\
+    }\n\
+\n\
+    if (element.style.getPropertyValue(propertyName) != propertyValue ||\n\
+        element.style.getPropertyPriority(propertyName) != \"important\")\n\
+      element.style.setProperty(propertyName, propertyValue, \"important\");\n\
+  }\n\
+\n\
+  doHide();\n\
+\n\
+  new MutationObserver(doHide).observe(\n\
+    element, {\n\
+      attributes: true,\n\
+      attributeFilter: [\"style\"]\n\
+    }\n\
+  );\n\
+}";
+
+  ss <<
+"\n\
+ var selector = \"[src$='" + relative_url + "'], [srcset$='" + relative_url + "']\";\n\
+ //console.log('blocked element selector: ' + selector);\n\
+ var elements = document.querySelectorAll(selector);\n\
+ if (elements.length > 0) {\n\
+   //console.log('found ' + elements.length + ' elemhide blocked elements for selector ' + selector);\n\
+   for (var e=0; e<elements.length; e++) {\n\
+     var element = elements[e];\n\
+     if (element.src == '" + url + "')\n\
+       hideElement(element);\n\
+   }\n\
+ }\n\
+console.log('finished elemhiding blocked resources');";
+
+  return ss.str();
+}
+
+void elemhideBlockedResource(
+    const base::Callback<content::WebContents*(void)>& wc_getter,
+    const base::Callback<int(void)>& id_getter,
+    const std::string& url,
+    const std::string& relative_url) {
+    DCHECK_CURRENTLY_ON(BrowserThread::UI);
+
+    content::WebContents* webContents = wc_getter.Run();
+    if (!webContents) {
+      LOG(ERROR) << "Adblock: elemhideBlockedResource no webContents";
+      return;
+    }
+
+    int frameTreeNodeId = id_getter.Run();
+    if (frameTreeNodeId < 0) {
+      LOG(ERROR) << "Adblock: negative frameTreeNodeId";
+      return;
+    }
+    LOG(WARNING) << "Adblock: frameTreeNodeId=" << frameTreeNodeId;
+
+    std::string js = generateBlockedElemhideJavaScript(url, relative_url);
+    //LOG(WARNING) << "Adblock: generated elemhide js for blocked url\n" << js;
+
+    // run JS
+    content::RenderFrameHost* frameHost = (frameTreeNodeId
+      ? webContents->UnsafeFindFrameByFrameTreeNodeId(frameTreeNodeId)
+      : webContents->GetMainFrame());
+
+    if (frameHost) {
+      // elemhide resource by elemehide rules
+      frameHost->ExecuteJavaScriptInIsolatedWorld(
```

Comment:
BTW, sorry I didn't look into the code of chromium. Does it execute the script completely isolated from the page or it can be clashed with the script of the page?

Reply:
it runs in isolated world (the same we use for element hiding for element hiding rules)

Reply:
Is the "world" new for each invocation or it's shared among several invocations?

Reply:
does it make sense?

i believe it's shared between invocations, but we can easily check it by calling 2 scripts with the same isolated world id in a row and let's say 1st creates variable and 2nd reads it.

https://developer.chrome.com/extensions/content_scripts#isolated_world

Reply:
Could you try it? If it's this way then we should re-use functions or wrap each injected JS by anonymous JS functions.

Reply:
it looks like not very significant improvement and could be done in separate task

Reply:
If there are at least two blocked requests then starting from the second one that JS will not work because one may not declare variables several times in JS using `let` and `const`.

Reply:
how can you explain that there are 3 blocked images on 50statesclassifieds.com and all of them are elemhidden? see screenshots at https://gitlab.com/eyeo/adblockplus/chromium/issues/12

Reply:
Can it be that for each frame there is a separate "world"?

Reply:
@xxxz  it seems you're right. testing on theverge.com:
> 08-29 17:11:29.113 22920-22920 I/chromium: [INFO:CONSOLE(1)] "Uncaught SyntaxError: Identifier 'selector' has already been declared", source: https://cdn.krxd.net/partnerjs/xdi/proxy.3d2100fd7107262ecb55ce6847f01fa5.html#!kxcid=JImcjrMY&kxt=https%3A%2F%2Fcdn.krxd.net&kxcl=cdn&kxp= (1)